File: 1_easy/02601_02800/02614_prime_in_diagonal/solution.cpp
class Solution {
    bool is_prime(int num) {
        if (num <= 1) {
            return false;
        }

        if (num == 2 || num == 3 || num == 5) {
            return true;
        }

        if (num % 2 == 0 || num % 3 == 0 || num % 5 == 0) {
            return false;
        }

        int sq = sqrt(num);
        for (int i{7}; i <= sq; i += 2) {
            if (num % i == 0) {
                return false;
            }
        }

        return true;
    }

public:
    int diagonalPrime(vector<vector<int>>& nums) {
        int n = nums.size();
        int max = 0;
        for(int i{0}; i < n; ++i) {
            if (max < nums[i][i] && is_prime(nums[i][i])) {
                max = nums[i][i];
            }
            if (max < nums[i][n - 1 - i] && is_prime(nums[i][n - 1 - i])) {
                max = nums[i][n - 1 - i];
            }
        }

        return max;
    }
};

File: 1_easy/02601_02800/02614_prime_in_diagonal/solution.py
class Solution:
    def is_prime(self, num):
        if num <= 1:
            return False

        if num == 2 or num == 3 or num == 5:
            return True

        if num % 2 == 0 or num % 3 == 0 or num % 5 == 0:
            return False

        sq = int(sqrt(num))
        for i in range(7, sq + 1, 2):
            if num % i == 0:
                return False

        return True

    def diagonalPrime(self, nums: List[List[int]]) -> int:
        n = len(nums)
        max_num = 0
        for i in range(n):

            if max_num < nums[i][i] and self.is_prime(nums[i][i]):
                max_num = nums[i][i]

            if max_num < nums[i][n - 1 - i] and self.is_prime(nums[i][n - 1 - i]):
                max_num = nums[i][n - 1 - i]

        return max_num
